### PR TITLE
Do not unify dereferences of shared borrows in GVN

### LINKED ISSUE #130853: Miscompile in the GVN transform
I tried this code:

```rust
unsafe fn src(x: &&u8) -> bool {
    let y = **x;
    unknown();
    **x == y
}

static mut SUSSY: *mut u8 = core::ptr::null_mut();

#[inline(never)]
unsafe fn unknown() {
    *SUSSY = 1;
}

fn main() {
    let mut s = 0;
    unsafe {
        SUSSY = core::ptr::addr_of_mut!(s);
        println!("{}", src(&*core::ptr::addr_of!(SUSSY).cast::<&u8>()));
    }
}

```

I expected to see this happen:

This should print `false`, as I believe this is DB under both Stacked and Tree borrows(according to MIRI).

Instead, this happened:

It returns `false` in Debug mode, and the GVN MIR pass makes `src()` unconditionally return `true` in Release mode.

```sh
$ cargo miri run
Compiling sus v0.1.0 (/Users/jwong3/test/sus)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.17s
     Running `/Users/jwong3/.rustup/toolchains/nightly-aarch64-apple-darwin/bin/cargo-miri runner target/miri/aarch64-apple-darwin/debug/sus`
false
```

```sh
$ cargo run -r
   Compiling sus v0.1.0 (/Users/jwong3/test/sus)
    Finished `release` profile [optimized] target(s) in 0.31s
     Running `target/release/sus`
true
```

### Meta

Here's the MIR before the GVN pass:

```mir
// MIR for `src` before GVN

fn src(_1: &&u8) -> bool {
    debug x => _1;
    let mut _0: bool;
    let _2: u8;
    let _3: ();
    let mut _4: u8;
    let mut _5: u8;
    let mut _6: &u8;
    let mut _7: &u8;
    scope 1 {
        debug y => _2;
    }

    bb0: {
        StorageLive(_2);
        _6 = deref_copy (*_1);
        _2 = copy (*_6);
        _3 = unknown() -> [return: bb1, unwind continue];
    }

    bb1: {
        StorageLive(_4);
        _7 = deref_copy (*_1);
        _4 = copy (*_7);
        StorageLive(_5);
        _5 = copy _2;
        _0 = Eq(move _4, move _5);
        StorageDead(_5);
        StorageDead(_4);
        StorageDead(_2);
        return;
    }
}
```

<details>

<summary>After</summary>

```mir
// MIR for `src` after GVN

fn src(_1: &&u8) -> bool {
    debug x => _1;
    let mut _0: bool;
    let _2: u8;
    let _3: ();
    let mut _4: u8;
    let mut _5: u8;
    let mut _6: &u8;
    let mut _7: &u8;
    scope 1 {
        debug y => _2;
    }

    bb0: {
        nop;
        _6 = copy (*_1);
        _2 = copy (*_6);
        _3 = unknown() -> [return: bb1, unwind continue];
    }

    bb1: {
        StorageLive(_4);
        _7 = copy _6;
        _4 = copy _2;
        StorageLive(_5);
        _5 = copy _2;
        _0 = const true;
        StorageDead(_5);
        StorageDead(_4);
        nop;
        return;
    }
}
```

</details>

It would be justified to make `src()` return `true` if `_6` was dereferenced again in `bb1`, however, the write in `unknown()` shouldn't invalidate the actual pointer stored in `_1` if my understanding of Stacked Borrows is correct.

This is present in both Stable and Nightly.

### PR DESCRIPTION
**Pull request by cjgillot**
The pass as written does not differentiate single deref or nested derefs, so I'm gating the entire unification on `-Zunsound-mir-opts`. This has the unfortunate effect of preventing reading from statics.

Fixes https://github.com/rust-lang/rust/issues/130853

r? @RalfJung

### CODE REVIEW

**Comment by DianQK** [comment]
Do you think it makes sense to invalidate all known `Deref`  when encountering with a dereference assignment or a function call? If so, then r=me.
Context: I'm trying to fix #128299, so I still need the `Deref` fact.

**Comment by RalfJung** [comment]
Thanks!
Can you add the example from https://github.com/rust-lang/rust/issues/130853 as a test?

File: tests/mir-opt/pre-codegen/deref_nested_borrows.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// EMIT_MIR_FOR_EACH_PANIC_STRATEGY
+
+fn src(x: &&u8) -> bool {
+    // CHECK-LABEL: fn src(
+    // CHECK-NOT: _0 = const true;
+    // CHECK: _0 = Eq({{.*}}, {{.*}});
+    // CHECK-NOT: _0 = const true;
+    let y = **x;
+    unsafe { unknown() };
```
Comment (DianQK) [inline tests/mir-opt/pre-codegen/deref_nested_borrows.rs:9]: Here~ @RalfJung

Comment (RalfJung) [reply tests/mir-opt/pre-codegen/deref_nested_borrows.rs:9]: Ah, that's drowning in all the other diff noise for tests changing their output.^^

I find mir-opt diffs completely unreadable so I largely skip them...

File: tests/mir-opt/pre-codegen/deref_nested_borrows.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// EMIT_MIR_FOR_EACH_PANIC_STRATEGY
```
Comment (RalfJung) [inline tests/mir-opt/pre-codegen/deref_nested_borrows.rs:1]:
```suggestion
//! Regression test for <https://github.com/rust-lang/rust/issues/130853>
// EMIT_MIR_FOR_EACH_PANIC_STRATEGY
```

**Comment by RalfJung** [comment]
Okay, the code diff as well as the new test seem right to me. I have absolutely no idea what to do with the remaining diff.^^

File: tests/mir-opt/const_prop/slice_len.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ fn main() {
     // CHECK-LABEL: fn main(
     // CHECK: debug a => [[a:_.*]];
     // CHECK: [[slice:_.*]] = copy {{.*}} as &[u32] (PointerCoercion(Unsize, AsCast));
-    // CHECK: assert(const true,
-    // CHECK: [[a]] = const 2_u32;
+    // CHECK-NOT: assert(const true,
+    // CHECK-NOT: [[a]] = const 2_u32;
```
Comment (RalfJung) [inline tests/mir-opt/const_prop/slice_len.rs:12]: Seems odd to have tests asserting that an optimization does *not* happen. Shouldn't we instead keep the old assertions but comment them out?

Comment (klensy) [reply tests/mir-opt/const_prop/slice_len.rs:12]: There is `COM:` for that,

**Comment by RalfJung** [comment]
Since @cjgillot seems to be busy I made an updated version of this PR: https://github.com/rust-lang/rust/pull/133474

**Comment by RalfJung** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/133474

**Pull request closed without merging**